### ISSUE TITLE
[Bug] Unable to open non-webp original images with extension

### ISSUE THREAD

**Issue by seth-rah** [furyutei/twOpenOriginalImage#28]
## Summary

Twitter changed something in its thumbnail view to now serve WebP images on the thumbnail view.

When opening the tweet image without the Chrome extension you are provided with a file using a different file format than that shown in the thumbnail view.

This mismatch is causing link generation to assume the thumbnail image is of the WebP format instead of being able to grab the original file extension.

-------

TL:DR - Thumbnail image WebP, original image different format, extension tries to show original as a WebP file, fails.

**Comment by seth-rah** [comment]
trying to find which function handles generating the link

I have a dirty workaround I could make a branch of, but I'm having trouble finding the correct place to put it

![image](https://github.com/furyutei/twOpenOriginalImage/assets/2222634/bd70a9c1-4083-4263-81eb-6eb7cf037ed3)

I was able to find the correct file extensions with some debug code but that didn't seem to work in the get_img_extension function when overwritten. So I'm assuming that the entire link is generated elsewhere

**Comment by seth-rah** [comment]
So thanks to dear old Musky not giving us a way to retrieve the original extension without doing a request, this option isn't very feasible...

Opening single tweets it's fine, but once you open your timeline and start scrolling, you effectively DDOS twitter and make your client die along the way.

This effectively restores functionality of the extension, but it's not worth the downsides.

```js
function get_url_info(url) {
  // For the sake of network requests, this workaround will only support jpg and png
  const extension_list = ['jpg', 'png'];
  let found_extension = null; // initialize variable to store extension value

  for (const extension of extension_list) {
    if (found_extension) break; // skip if extension already found
    const modifiedUrl = url.replace(/^(.*)(\?format=[^&]+.*)$/, `$1?format=${extension}&name=orig`);

    const request = new XMLHttpRequest();
    try {
      request.open('HEAD', modifiedUrl, false); // Make the request synchronous
      request.send(null);
    } catch (error) {}

    if (request.status === 200) {
      found_extension = extension; // store extension value when found
      url = url.replace(/^(.*?)(\.\w+)(.*?format=)([^&]+)(\&.*name=)(\w+)(&.*)?$/, `$1$2$3${found_extension}$5$6`); // update the URL
      String(url)
    }
  }

  // [注意] url は https?:// 以外（画像ファイル名など）の場合あり（※その場合はnew URL(url)とするとエラー発生）
  var url_parts = url.split('?'),
    query_map = {},
    url_info = { base_url: url_parts[0], query_map: query_map };

  if (url_parts.length < 2) {
    return url_info;
  }

  url_parts[1].split('&').forEach(function (query_part) {
    var parts = query_part.split('=');

    query_map[parts[0]] = parts.length < 2 ? '' : parts[1];
  });

  return url_info;
} // end of get_url_info()
```  

This could work if the extension had a way of knowing which URL's it interacted with before, but right now it tries to refetch everything, even when it has found a value before.

This leads to your console looking like this, and it will keep going, even though it knows this is a png already.

![image](https://github.com/furyutei/twOpenOriginalImage/assets/2222634/b13671e3-44cb-4af2-a644-97a9c87e4094)

Maybe there is a way to utilise the `twOpenOriginalImage_touched` tag so that it knows when it already has the right address, but the only real way to store that info would be to always replace the timeline images with the original one or creating a db.

**Comment by thiagojramos** [comment]
Here's a possible solution.

To work around this issue, we need to come up with a solution that minimizes network requests and is able to efficiently store and retrieve already checked image file extensions. Here's a possible approach:

1. **Caching**: We can use the browser's local storage (like localStorage or IndexedDB) to store the extensions of already checked image files. This will allow us to check if we already know the file extension of an image before making a network request. However, be aware that local storage has size limits and that the data may be deleted by the browser under certain circumstances.

2. **Batch network requests**: Instead of making a network request for each image, we can try to make batch requests. This might be a bit more complex as we would need to manage a queue of images whose file extensions still need to be checked.

3. **Fallback to WebP**: If we cannot determine the file extension of an image, we could assume it's WebP. While this isn't ideal, it's better than not displaying the image at all.

Here's some pseudocode that demonstrates the idea:

```javascript
// Cache storage for file extensions
let fileExtensionCache = new Map();

function getFileExtension(url) {
  // Check if the file extension is already cached
  if (fileExtensionCache.has(url)) {
    return fileExtensionCache.get(url);
  }

  // Otherwise, make a network request to determine the file extension
  // ...

  // If we were able to determine the file extension, cache it
  if (extension) {
    fileExtensionCache.set(url, extension);
  }

  // Return the file extension (or null if we couldn't determine it)
  return extension;
}
```

Note: this is a simplified example and you'll need to adapt it according to your specific needs. Also, always keep in mind the limitations and restrictions of the browser environment when working with local storage and network requests.

**Comment by seth-rah** [comment]
That might work, but I feel like the extension might have a bit of an infrastructural issue with constantly shotgunning these requests.

`get_url_info` seems to be what most functions rely on to get image data.

opening a link runs through all these calls again from the timeline as well as re-processing the link you select.

I don't mind too much doing processing on a link as you select it as that's relatively lightweight. It's more now that it seems to be that while scrolling it tries to handle data it's already replaced or isn't relevant anymore.

![image](https://github.com/furyutei/twOpenOriginalImage/assets/2222634/efe06ada-bf35-49a7-8494-3c42ede944ef)

I don't know which function is responsible, but if I can find what's constantly invoking get_url_info while scrolling then it might be possible to avoid needing to create a map of files that have already been processed, as the data is already present in the HTML

**Comment by thiagojramos** [comment]
It seems that the problem persists because the `get_url_info` function is constantly being called as the user scrolls through the page, reprocessing data that has already been replaced or is no longer relevant. This appears to be causing the overload of network requests you mentioned.

The `get_url_info` function is used by the extension to get information about an image, including its URL and its file extension. It's called whenever a new image is loaded onto the page, which happens constantly as you scroll through Twitter. The constant invocation of this function is likely what's causing the issue.

In the image he provided, he highlights the `twOpenOriginalImage_touched` class, which appears to be added to an image after its original file extension has been determined and the image's URL has been updated.

He's suggesting that instead of creating a map of files that have already been processed (as I proposed in my earlier solution), we could try to use the information already present in the page's HTML. In other words, we could check for the presence of the `twOpenOriginalImage_touched` class on an image before attempting to determine its original file extension.

Here's an example of how you might do this:

```javascript
function getFileExtension(element) {
  // Check if the image has already been processed
  if (element.classList.contains('twOpenOriginalImage_touched')) {
    // The image has already been processed, so we can extract the file extension from the image's URL
    let url = element.getAttribute('src');
    let extension = url.split('?')[0].split('.').pop();

    return extension;
  }

  // Otherwise, make a network request to determine the file extension
  // ...
}
```

However, this assumes that the image's URL in the page's HTML is always updated to reflect the image's original file extension. If this isn't the case, this approach won't work.

In any case, I believe the definitive solution to this problem would be to refactor the extension to avoid the constant re-invocation of the `get_url_info` function. This would likely require a deeper understanding of the extension's code and perhaps significant restructuring of it.

**Comment by seth-rah** [comment]
I do agree that this extension needs a refactor, it's got some legacy code. Just adding one async call turns this into a big pile of spaghetti. But for now I'm just trying to apply a band-aid rather than trying to refactor the extension as that would require a bit more of my time than I'd be willing to spend on this.

So what I see is that even if you don't replace the original URL, the image still gets the touched class added to it.

Opening an image still runs the get_url function and will allow the original to be pulled while the timeline has the webp file displayed.

![image](https://github.com/furyutei/twOpenOriginalImage/assets/2222634/50174446-0cbf-47cc-9da4-50c72b9199e2)

![image](https://github.com/furyutei/twOpenOriginalImage/assets/2222634/dc803b8c-5a8e-4d54-b7bd-f13f44c4ca13)

I've found where the logic invokes that handles get_url_info, so I'm trying to see what I can do to apply the check for the tag

It seems the tag is removed under certain conditions as well, so I'm trying to see what causes that and if it's needed

**Comment by thiagojramos** [comment]
Well, it seems like you're making some progress, but I agree with you, this extension could use a decent refactor. But let's focus on that band-aid for now.

The behavior you're describing about the "touched" class being added to images regardless of URL replacement is intriguing. It suggests that the extension is processing all images, regardless of whether it needs to retrieve the original file format.

Opening an image still runs the get_url_info function and allows the original to be pulled while the timeline displays the WebP file. However, you mention that your solution doesn't actually fix the replace timeline image, which might be something we need to address.

Looking at the settings you provided in your screenshot, I see that you have the 'Replace images on timeline with original-sized ones' option enabled. This option could be the reason why the extension is trying to replace every image on the timeline, leading to the excessive network requests.

You might want to consider disabling this option as a temporary solution, while we figure out a more permanent fix.

As for the HTML snippet, it seems that the twOpenOriginalImage_touched class is added to the image tag, but not to the div that contains the background-image CSS property. This could potentially explain why the extension is having trouble finding the original file extension - it's looking at the img tag, but the original file extension might be stored in the div's background-image URL.

You've mentioned that you've found where the logic invokes get_url_info and that you're trying to apply the check for the tag. That's a good starting point.

As for the behavior you noticed with the get_url_info function being run when an image is opened, that's an interesting find. It's possible that this behavior is contributing to the excessive network requests. If the function is being run every time an image is opened, it could be repeatedly making network requests for images that have already been processed.

It's also interesting that the tag is removed under certain conditions. Understanding these conditions might help us figure out why the extension is making so many network requests.

**Comment by seth-rah** [comment]
I've pushed a band-aid for now to a fork.

Created a pull request, but more work is definitely needed here.

For now this will make most images work in the overlay while some don't get added correctly (uncertain on the cause, but it only started after my changes so it's new)

For replacing thumbnail images with the original image, I think this should do the job for most scenarios. So being able to save images from thumbnails is at least one benefit.

https://github.com/furyutei/twOpenOriginalImage/pull/29

and the fork if you want to mess around. It's late here so I think I'll call it a night. https://github.com/furyutei/twOpenOriginalImage/compare/master...seth-rah:twOpenOriginalImage:query-image-address

I don't really see myself revisiting this any time soon though unless it's a minor change

**Comment by seth-rah** [comment]
In an ideal world Musk just has the original image format available somewhere in the html so we don't need to query different extensions over the network. I understand why webp was pushed for, but it hurts extensions like this.

**Comment by Marcos875433** [comment]
For now I just modified the [Line 779](https://github.com/furyutei/twOpenOriginalImage/blob/master/src/js/twOpenOriginalImage.user.js#L779) and replaced the `RegExp.$1` with `'jpg'`
```javascript
function get_img_url( img_url, kind, old_format ) {
    img_url = normalize_img_url( img_url );
    
    if ( old_format ) {
        if ( ! kind ) {
            kind = '';
        }
        else {
            if ( kind.search( ':' ) != 0 ) {
                kind = ':' + kind;
            }
        }
        img_url = img_url.replace( /:\w*$/, '' ) + kind;
    }
    else {
        if ( ! kind ) {
            kind = 'orig';
        }
        kind = kind.replace( /:/g, '' );
        
        if ( ! /:\w*$/.test( img_url ) ) {
            img_url += ':' + kind;
        }
        
        img_url = img_url.replace( /\.([^.]+):\w*$/, '' ) + '?format=' + 'jpg' + '&name=' + kind;
    }
    
    return img_url;
} // end of get_img_url()
```
It seems to work for now, it just takes a little bit more time (at least for me) to load some of the images, good if you just wanna look at the original image (you can also replace it with `'png'` but since from what I've seen png doesnt have `name=orig` so instead you can replace the line with):
```javascript
img_url = img_url.replace( /\.([^.]+):\w*$/, '' ) + '.png';
```
Though, it will take a decent amount of time for some high resolution images to load in the case of pngs.
And also, you can do the same thing for jpg:
```javascript
img_url = img_url.replace( /\.([^.]+):\w*$/, '' ) + '.jpg';
```
I used the first solution just to stay as close as possible to the original code.

And btw, from all the images I tested with, all of them seemed to have a png, webp, and jpg version, so maybe all (or at least good amount) of the images uploaded have png, webp, and jpg version, but regardless, I couldn't find a way to figure out the original image format, I even looked through the cached images, and they didnt seem to have a different format other than the one that the url shows, also, downloaded some image in webp format, analyzed the [metadata](https://www.metadata2go.com/view-metadata), and it seemed to actually be type webp, so there seems to not be at least an intuitive way of retrieving the original format.

**Comment by seth-rah** [comment]
So the issue with that is you won't be able to pull the original file resolution as well, which will effectively make the use of this extension redundant

Example:

https://pbs.twimg.com/media/FHScljOXoAMT87a?format=jpg

https://pbs.twimg.com/media/FHScljOXoAMT87a?format=jpg&name=orig

This is what makes this entire situation so annoying right now, is that the only way to get the original extension is either by opening the tweet and loading the image, or doing some requests

**Comment by seth-rah** [comment]
Pushed another small update to not ignore images smaller than 240px, I now filter on timeline images via the aria-label. (works in conversations too)

problem is that I don't know if this is different for JP.

**Comment by seth-rah** [comment]
Fixed https://github.com/furyutei/twOpenOriginalImage/commit/442b7ea12322a8d9811623aa720374401b3a58ce
